client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -212,66 +222,50 @@ private Response performRequest(final NodeTuple<Iterator<Node>> nodeTuple,
         } catch(Exception e) {
             RequestLogger.logFailedRequest(logger, request.httpRequest, context.node, e);
             onFailure(context.node);
-            Exception cause = unwrapExecutionException(e);
+            Exception cause = extractAndWrapCause(e);
             addSuppressedException(previousException, cause);
             if (nodeTuple.nodes.hasNext()) {
                 return performRequest(nodeTuple, request, cause);
             }
             if (cause instanceof IOException) {
-                throw (IOException)cause;
+                throw (IOException) cause;
             }
             if (cause instanceof RuntimeException) {
-                throw (RuntimeException)cause;
+                throw (RuntimeException) cause;
             }
-            throw new RuntimeException(cause);
+            throw new IllegalStateException("cause must be either RuntimeException or IOException", cause);
```

Comment:
Maybe "unexpected exception type so wrapping into an expected one to prevent even more chaos"?